### PR TITLE
Initial hw2 for David Stanley (DavidMS4)

### PR DESCRIPTION
**Pull request by DavidMSCode**

### CODE REVIEW

**Comment by npnike10** [comment]
Hello, I'll review your code.

File: Trained.gif
Comment (ravadap) [inline Trained.gif:1]: Trained agent gif looks good

File: figures/test_discreteaction_pendulum.png
Comment (ravadap) [inline figures/test_discreteaction_pendulum.png:1]: Trajectory plot looks good

Missing learning curve, policy plot, state-value plot, and ablation study

File: train_discreteaction_pendulum.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        return random.sample(self.memory, batch_size)
+
+    def __len__(self):
+        """Return number of stored memories"""
+        return len(self.memory)
+
+# Create the Q network class
+
+
+class DQN(nn.Module):
```
Comment (ravadap) [inline train_discreteaction_pendulum.py:50]: Correct neural network initialization

File: train_discreteaction_pendulum.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ALF = 1e-4  # learning rate of optimizer
+C = 200     #frequency to reset the target network
+
+actions = range(env.num_actions)
+n_actions = len(actions)
+
+s = env.reset()
+n_obs = len(s)
+
+Q_net = DQN(n_obs, n_actions)  #create a network for the Q function
```
Comment (ravadap) [inline train_discreteaction_pendulum.py]: Neural network creation and initialization looks good

File: train_discreteaction_pendulum.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+
+def select_action(state, explore=True):
+    if not torch.is_tensor(state):
+            state = torch.tensor(state, dtype=torch.float32,
+                         device=device).unsqueeze(0)
+            
+    global steps #get steps for exploration decay
+    sample = random.random() #get a float from 0 to 1
+    eps = EPSF+(EPS0-EPSF)*math.exp(-1*steps/DECAY) #calculate epsilon for current step
```
Comment (ravadap) [inline train_discreteaction_pendulum.py]: Good use of decaying epsilon (not required for hw assignment right?)

File: train_discreteaction_pendulum.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    # Loss
+    optimizer.zero_grad()
+    criterion = nn.SmoothL1Loss()
+    loss = criterion(state_action_values,
+                     expected_state_action_values.unsqueeze(1))
+
+    # Optimize the model
+    loss.backward()
+    # In-place gradient clipping
+    torch.nn.utils.clip_grad_value_(Q_net.parameters(), 100)
```
Comment (ravadap) [inline train_discreteaction_pendulum.py:149]: I was unsure about this if the gradients need to be clipped during or after the back propagation and if that made a difference.

File: Trained.gif
Comment (ravadap) [inline Trained.gif:1]: Algorithms and implementation overall looks good
Only suggestion I would make is making a training function to make the ablation study easier

Comment (DavidMSCode) [reply Trained.gif:1]: thanks

File: train_discreteaction_pendulum.py (hunk 4 above)
Comment (DavidMSCode) [reply train_discreteaction_pendulum.py:149]: I think you're right they should be before the back prop. Though I'm guessing the clipping after is just making them be clipped for the next step and still prevents the exploding grads

File: train_discreteaction_pendulum.py (hunk 3 above)
Comment (DavidMSCode) [reply train_discreteaction_pendulum.py]: not required as far as I know

File: figures/test_discreteaction_pendulum.png
Comment (DavidMSCode) [reply figures/test_discreteaction_pendulum.png:1]: Plots always take me the longest...

**Review by ravadap — COMMENTED** [review]
Looks good :)

**Comment by DavidMSCode** [comment]
> Looks good :)

I don't know how I requested another review. I think I clicked something wrong in vscode but thank you.
